### PR TITLE
Richard/dev

### PR DESCRIPTION
**Pull request by r-barnes**
Adds cmake build capability and outputs files to a common directory specified by params files.

### CODE REVIEW

File: src/simul.c [hunk 1]
```diff
@@ -1561,6 +1562,7 @@ void simul_dump() {
       if (dump_delay_png && (time_threshold == time_threshold_dpng)) {
         //dump PNG file
         sprintf(name, "%s%05d_t0%s.png", MOD_NAME, cpt_dump, str);
+        output_path_noext(name);
```
Comment (r-barnes) [inline src/simul.c]: @defaziogiancarlo : Does this conflict with your work?

Comment (r-barnes) [reply src/simul.c]: It outputs images and .csp.gz files to the directory specified by the params file.

**Comment by kellykochanski** [comment]
Rebased to go into next release, rescal-release-1.1.

**Comment by r-barnes** [comment]
Updated to merge into the 1.1 release branch.

Pulled LICENSE out of docs, per requirements by Github and JOSS.

Recommend the same with NEWS, if you're going to have it.

**Comment by kellykochanski** [comment]
The README installation instructions suggest using a new "build" directory, but if I try to use make in that directory, I get the following error:
```bash
[kellykochanski]$ cmake -Wno-dev -DCMAKE_BUILD_TYPE=Releas
CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Error: The source directory "/g/g14/kochansk/rescal-tmp/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
```

**Comment by kellykochanski** [comment]
If I treat the above as a directory structure problem and run cmake directly in the rescal-snow directory, I get a series of `-Werror` warnings for C++ style comments such as:
```bash
/g/g14/kochansk/rescal-tmp/src/cells.h:48:21: error: (this will be reported only once per input file) [-Werror]
/g/g14/kochansk/rescal-tmp/src/csp2bin.c:37:47: error: C++ style comments are not allowed in ISO C90 [-Werror]
 int32_t H = 0, L = 0, D = 0, HL = 0, HLD = 0; // les dimensions de la terre
```

File: src/simul.c (hunk 1 above)
Comment (defaziogiancarlo) [reply src/simul.c]: In cellspace.py there's an assumption that the .csp files are in the same directory as the run script. That's an easy fix, I'll change it.

**Comment by r-barnes** [comment]
You run `cmake -Wno-dev -DCMAKE_BUILD_TYPE=Release ..` so that the build doesn't contaminate the source files.

**Comment by r-barnes** [comment]
I've set the C standard now to 99, so those errors should drop.

**Comment by kellykochanski** [comment]
Great. The build runs smoothly now.

**Comment by kellykochanski** [comment]
I'm hitting one new `-Werror`:

```bash
/g/g14/kochansk/rescal-tmp/src/entry.c:210:11: error: variable 'nRetVal' set but not used [-Werror=unused-but-set-variable]
   int32_t nRetVal;
           ^
cc1: all warnings being treated as errors
```

**Comment by r-barnes** [comment]
That's because release mode disables asserts, so the line `assert(nRetVal==0);` which used the variable doesn't get compiled. The solution is to actually check the return and close the program if there was an issue.

**Comment by r-barnes** [comment]
(All of these problems were already present in the code doing Bob Knows What, we're just using compile-time checks to find them now.)

**Comment by kellykochanski** [comment]
The build procedure ppears to write to the `rescal-snow` main directory, leaving the 'build' directory empty. 
```bash
[kellykochanski:build]$ cmake -Wno-dev -DCMAKE_BUILD_TYPE=Release ..
-- Symbol prefix: 
-- Configuring done
-- Generating done
-- Build files have been written to: $HOME/rescal-snow
```
This causes errors in `scripts` files, which look for the executables `../build/rescal` and `../build/genesis` instead of `../rescal` and `../genesis`

**Comment by r-barnes** [comment]
That's unusual. I suggest scrapping the build directory (`rm -rf build/`) and reseting the branch: (`git reset --hard` and `git clean -f`) to eliminate temporaries. But only do the previous if you don't have files that are not checked in that you want to keep.

May be an artefact of you having built outside of the build directory.

**Comment by r-barnes** [comment]
Can you also do a `cmake --version`? I want to eliminate some of those cmake dev errors but want to keep things working for you.

**Comment by kellykochanski** [comment]
`cmake version 3.14.5`

**Comment by kellykochanski** [comment]
After reset, I'm getting a new `-Werror`:

```bash
/g/g14/kochansk/rescal-tmp/src/transitions.c: In function 'dump_transitions':
/g/g14/kochansk/rescal-tmp/src/transitions.c:704:3: error: this 'for' clause does not guard... [-Werror=misleading-indentation]
   for(i=0; i<nb_liens; i++)
   ^~~
/g/g14/kochansk/rescal-tmp/src/transitions.c:706:5: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the 'for'
     output_write("TRANSITIONS", current_output);
     ^~~~~~~~~~~~
```

**Comment by kellykochanski** [comment]
Output is being written to new directories `scripts/??d` and `scripts/?,d`. These seem like odd defaults.

**Comment by kellykochanski** [comment]
Examples all produce expected output (except for odd directory names). Nice work.

**Comment by r-barnes** [comment]
Can you do an ls that shows the directories and where they are in the file tree?

**Comment by r-barnes** [comment]
I can't find any unicode.

**Comment by kellykochanski** [comment]
The oddly named directories live wherever rescal tries to put output; for the example cases, that's mostly `rescal-snow/scripts/??d`; for the parallel test cases, it's `rescal-snow/test_parallel_runs/tauMin0_lambdaI0.01/??c` etc.

Removing the directories reveals various names, e.g:
```bash
rm: cannot remove 'test_parallel_runs//tauMin0_lambdaI0.001/\360\274c': Is a directory
```

**Comment by kellykochanski** [comment]
The output write function in simul.c appears to append `\0` before the filename, instead of a directory name:

```bash
1703 void output_write(char *output_filename, char *output_content){
1704   // Write output string output_content
1705   // to file output_directory/output_name.log
1706   // (This opens and closes the file every time it is called.
1707   //  Inefficient if called often for small blocks of text.
1708   //  Currently this happens in numerous functions.)
1709   FILE *fp;
1710   char path[256] = {'\0'};
1711 
1712   if (strlen(output_filename) + strlen(output_content) > 250){
1713     ErrPrintf("Unexpectedly long file/directory name");} 
1714     
1715   strcat(path, output_filename);
1716   output_path(path);
1717   
1718   fp = fopen(path, "a");
1719   if (! fp ) {
1720           ErrPrintf("ERROR: cannot open file: %s \n", path);
1721           exit(-1);
1722   }       
1723   fprintf(fp, "%s", output_content);
1724   fclose(fp); 
1725 } 
```

**Comment by r-barnes** [comment]
Oh, I thought we were talking about a build issue.

**Comment by r-barnes** [comment]
You never confirmed that we fixed this:
 > The build procedure ppears to write to the rescal-snow main directory, leaving the 'build' directory empty.
So I misinterpreted this:
> Output from first two runs is being written to new directories scripts/??d and scripts/?,d

**Comment by r-barnes** [comment]
I've reverted the commit that brought in that change, so this PR is now only cmake, png upgrade, and a number of other small fixes to documentation and licenses.

**Comment by kellykochanski** [comment]
Successful fix for naming problem - output directory is now named "out" not "??c".

**Comment by r-barnes** [comment]
Great! Do we have anything left to do on this PR?

**Comment by kellykochanski** [comment]
The commit you reverted did organize the output usefully, and could be useful in a future release if the naming problem were fixed.

**Pull request merged**